kollect-core/src/main/kotlin/kollect/Kollect.kt

```diff
@@ -0,0 +1,49 @@
+package kollect
+
+import arrow.TC
+import arrow.core.Tuple2
+import arrow.data.NonEmptyList
+import arrow.free.Free
+import arrow.higherkind
+
+typealias Kollect<A> = Free<ForKollectOp, A>
+
+abstract class NoStackTrace : Throwable() {
+    override fun fillInStackTrace(): Throwable = this
+}
+
+sealed class KollectError : NoStackTrace() {
+    abstract val env: Env
+}
+data class NotFound(override val env: Env, val request: KollectOne<Any, Any>) : KollectError()
+data class MissingIdentities(override val env: Env, val missing: Map<DataSourceName, List<Any>>) : KollectError()
+data class UnhandledException(override val env: Env, val err: Throwable) : KollectError()
+
+sealed class KollectRequest
+
+sealed class KollectQuery<I : Any, A> : KollectRequest() {
+    abstract fun dataSource(): DataSource<I, A>
+    abstract fun identities(): NonEmptyList<I>
+}
+
+/**
+ * Primitive operations in the Kollect Free monad.
+ */
+@higherkind
+interface KollectOp<A> : TC
```

Comment:
Does this need to be public API? Since Kotlin does not have `sealed interfaces` I would be inclined to make it internal to the module. But since`@higherkind` is currently not taking account visibility modifiers there is no way to do this at the moment.

Reply:
Should I create an issue in arrow?